File: samples/java/DeliveryDispatch/Server/Dispatch/src/main/java/systems/zlink/samples/deliverydispatch/server/dispatch/handlers/OfferDeliveryResultHandler.java
package systems.zlink.samples.deliverydispatch.server.dispatch.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkSendHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.dispatch.DeliveryOfferStore;
import systems.zlink.samples.deliverydispatch.server.dispatch.DeliveryOfferStore.DeliveryOffer;
import systems.zlink.samples.deliverydispatch.server.dispatch.DispatchWorker;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/**
 * The courier's decision, arriving as its own inbound message rather than as the reply to a request
 * nobody could have made. A decision naming an attempt other than the one on record came back after
 * the offer had already been reassigned, and is dropped — that check is what makes it safe never to
 * wait (common sample spec section 7.4).
 */
@ZLinkHandlerGroup(SampleNames.DispatchChannel)
public final class OfferDeliveryResultHandler
        implements ZLinkSendHandler<Messages.OfferDeliveryResultMsg> {
    private final DeliveryOfferStore offers;
    private final DispatchWorker worker;

    public OfferDeliveryResultHandler(DeliveryOfferStore offers, DispatchWorker worker) {
        this.offers = offers;
        this.worker = worker;
    }

    @Override
    public CompletionStage<Void> handle(
            Messages.OfferDeliveryResultMsg message, ZLinkMessageContext context) {
        // --8<-- [start:doc-dd-decision-settle]
        Optional<DeliveryOffer> offer = offers.settle(message.deliveryId(), message.attempt());
        if (offer.isEmpty()) {
            System.out.println(
                    "deliverydispatch-dispatch stale-decision-ignored delivery="
                            + message.deliveryId()
                            + " courier="
                            + message.courierId()
                            + " attempt="
                            + message.attempt());
            return CompletableFuture.completedFuture(null);
        }

        System.out.println(
                "deliverydispatch dispatch: decision delivery="
                        + message.deliveryId()
                        + " courier="
                        + message.courierId()
                        + " attempt="
                        + message.attempt()
                        + " accepted="
                        + message.accepted());
        return worker.settle(offer.get(), message.accepted(), message.reason());
        // --8<-- [end:doc-dd-decision-settle]
    }
}

File: samples/kotlin/DeliveryDispatch/Server/Dispatch/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/dispatch/handlers/OfferDeliveryResultHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.dispatch.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingSendHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.dispatch.DeliveryOfferStore
import systems.zlink.samples.kotlin.deliverydispatch.server.dispatch.DispatchWorker
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.OfferDeliveryResultMsg

/**
 * The courier's decision, arriving as its own inbound message rather than as the reply to a request
 * nobody could have made. A decision naming an attempt other than the one on record came back after
 * the offer had already been reassigned, and is dropped — that check is what makes it safe never to
 * wait (common sample spec section 7.4).
 */
@ZLinkHandlerGroup(SampleNames.DispatchChannel)
class OfferDeliveryResultHandler(
    private val offers: DeliveryOfferStore,
    private val worker: DispatchWorker,
) : ZLinkSuspendingSendHandler<OfferDeliveryResultMsg> {
    override suspend fun handle(message: OfferDeliveryResultMsg, context: ZLinkMessageContext) {
        // --8<-- [start:doc-dd-decision-settle]
        val offer = offers.settle(message.deliveryId, message.attempt)
        if (offer == null) {
            println(
                "deliverydispatch-dispatch stale-decision-ignored " +
                    "delivery=${message.deliveryId} courier=${message.courierId} " +
                    "attempt=${message.attempt}"
            )
            return
        }

        println(
            "deliverydispatch dispatch: decision delivery=${message.deliveryId} " +
                "courier=${message.courierId} attempt=${message.attempt} accepted=${message.accepted}"
        )
        worker.settle(offer, message.accepted, message.reason)
        // --8<-- [end:doc-dd-decision-settle]
    }
}
